File: Advent_of_Code/y2022/d16/p.rs
use std::collections::{BinaryHeap, HashMap, HashSet};
use std::env::args;

#[derive(Clone)]
struct Node {
    valve: String,
    time: i8,
    released: u16,
    visited: HashSet<String>,
    opened_all: bool,
}

impl Eq for Node {}

impl PartialEq for Node {
    fn eq(&self, other: &Self) -> bool {
        self.priority() == other.priority()
    }
}

impl Ord for Node {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.priority().cmp(&other.priority())
    }
}

impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Node {
    fn new(
        valve: String,
        time: i8,
        released: u16,
        visited: HashSet<String>,
        pressured: &HashSet<String>,
    ) -> Self {
        Self {
            valve,
            time,
            released,
            opened_all: pressured.is_subset(&visited),
            visited,
        }
    }

    fn explore(&mut self, d: &HashMap<String, (u8, Vec<String>)>) {
        if !self.visited.contains(&self.valve) {
            self.released += self.time as u16 * d[&self.valve].0 as u16;
            self.visited.insert(self.valve.clone());
            self.time -= 1;
        }
        self.time -= 1;
    }

    fn priority(&self) -> u16 {
        self.released
    }
}

fn main() {
    if args().len() != 2 {
        panic!("usage: cargo run --bin y2022d16 <input file>");
    }

    let lines: Vec<String> = std::fs::read_to_string(args().nth(1).unwrap())
        .unwrap()
        .lines()
        .map(|s| s.to_string())
        .collect();

    let mut d: HashMap<String, (u8, Vec<String>)> = HashMap::new();

    for l in lines {
        let l = l
            .chars()
            .filter(|c| c.is_ascii_uppercase() || c.is_numeric() || c.is_ascii_whitespace())
            .collect::<String>();
        let l = l
            .split(' ')
            .filter(|s| !s.is_empty())
            .skip(1)
            .collect::<Vec<&str>>();

        let mut v = Vec::new();
        for i in l.iter().skip(2) {
            v.push(i.to_string());
        }

        d.insert(l[0].to_string(), (l[1].parse().unwrap(), v));
    }

    let mut q: BinaryHeap<Node> = BinaryHeap::from(vec![Node::new(
        "AA".to_string(),
        30,
        0,
        HashSet::new(),
        &d.keys().cloned().collect(),
    )]);

    let mut max = 0;
    let mut i: usize = 0;

    while let Some(cur) = q.pop() {
        i += 1;
        if i > 1_000 {
            print!("{}\t{}\r", q.len(), cur.released);
            i = 0;
        }

        if cur.time <= 0 || cur.opened_all {
            if cur.released > max {
                max = cur.released;
                println!("\nnew max:\t{}", max)
            }
            continue;
        }
        for v in &d[&cur.valve].1 {
            let mut n = cur.clone();
            n.valve = v.clone();
            n.explore(&d);
            q.push(n);
        }
    }

    println!("{}", max);
}

File: advent_of_code/y2022/d16/p.py
#!/usr/bin/env python3

import heapq
import re
from os.path import dirname
from typing import Dict, List, Set, Tuple, Union

from aocd import get_data

DAY = int(re.sub(r"[^0-9]", "", dirname(__file__).split("/")[-1]))
YEAR = int(re.sub(r"[^0-9]", "", dirname(__file__).split("/")[-2]))
DATA: str = get_data(day=DAY, year=YEAR)

lines = DATA.splitlines()

d: Dict[str, Tuple[int, List[str]]] = {}
pressure: Set[str] = set()

for l in lines:
	s = re.sub("[(rate=};,]", "", l).split()
	d[s[1]] = (int(s[4]), s[9:])
	if d[s[1]][0] > 0:
		pressure.add(s[1])


class Node:
	def __init__(self, valve: str, time: int, released: int, visited: Set[str]):
		self.valve = valve
		self.time = time
		self.released = released
		self.visited = visited
		self.all_visited: bool = pressure.issubset(visited)

	def __lt__(self, other):
		return self.priority() > other.priority()

	def __repr__(self):
		return f"Node({self.valve}, {self.time}, {self.released}, {self.visited})"

	def priority(self) -> int:
		return (
			int(self.valve in self.visited) * -100_000
			+ self.released
			+ (30 - self.time) * 100_000
			+ int(self.all_visited) * 100_000_000
		)

	def explore(self):
		if self.valve not in self.visited:
			self.released += self.time * d[self.valve][0]
			self.visited.add(self.valve)
			self.time -= 1
		self.time -= 1


open = [Node("AA", 30, 0, set())]
heapq.heapify(open)
max_realease = 0

i = 0
while open:
	if i % 100_000 == 0:
		print(f"{i}\t{len(open)}", end="\r")
	i += 1
	i %= 100_000_000

	node = heapq.heappop(open)
	node.explore()

	if node.time <= 0 or node.all_visited:
		if node.released > max_realease:
			max_realease = node.released
			print(f"\nnew max:\t{max_realease}")
		continue

	for v in d[node.valve][1]:
		heapq.heappush(open, Node(v, node.time, node.released, node.visited.copy()))

print(f"part 1:\t{max_realease}")
